[feat] web-BE issue 관련 api 구현

j03y14 opened the pull request:
# web-BE issue 관련 api 구현
## 관련 이슈 및 위키
**관련 이슈** : #23 #24 #25 #26 #27 #28 #29 #30 
## 내용
### issue api 구현
- create issue
- read all issue
- update issue title
- add issue milestone
- delete issue milestone
- add issue label
- delete issue label
- update isopen
- add issue assignee
- delete issue assignee
- read all assignee

### 주의할 부분?
> 9346ac2

issue table에 추가하고 나서 issue_label table에도 추가를 해야하는데 둘 중 하나만 값이 넣어지는 것을 방지하기 위해서 transaction을 사용했습니다.

> 9749b92

issue를 먼저  milestone, user과 join 해서 가져온 후에 각 issue에 해당하는 label들을 가져와서 합친 후 반환합니다.

> 8693edc

labelArr의 label의 개수에 따라서 중첩 array를 만들어서 bulk insert를 하도록 했습니다.

## Why?
> 9749b92

writer, assignee까지 join 해서 값을 가지고 오지 않은 이유는 User를 join 할 때 writer와 assignee를 구분할 수 있는 방법을 모르기 때문에 따로 api를 만들었습니다.

j03y14 committed:
[feat] isnertIssue, insertIssueLabel query 추가

insertIssue 쿼리에서는 writer_id를 넣어야하기 때문에 내부 query를 사용해서 username으로 writer_id를 User table에서 가져와서 insert 해주었습니다.

insertIusseLabel은 bulk insert를 하기 위해서 VALUSE 이후에 ? 하나만 넣어주었습니다.

j03y14 committed:
[feat] issue model의 insert 메소드 추가

issue table에 추가하고 나서 issue_label table에도 추가를 해야하는데 둘 중 하나만 값이 넣어지는 것을 방지하기 위해서 transaction을 사용했습니다.

에러가 생기는 경우 응답코드 500번으로 에러를 throw해서 controller에서 처리하도록 했습니다.

j03y14 committed:
[feat] issue_label 모델, insert 함수 추가

labelArr의 label의 개수에 따라서 중첩 array를 만들어서 bulk insert를 하도록 했습니다.

j03y14 committed:
[feat] issue router 추가

issue router와 controller가 구현이 되었기 때문에 issue router의 주석을 해제했습니다.

j03y14 committed:
[feat] issue controller 추가

issue controller에서 create 메소드를 만들고 issueModel에게 insert를 하도록 issueModel의 insert 메소드를 호출하도록 했습니다.

j03y14 committed:
[feat] issue post ' /' router 추가

issue create까지 완료가 되어 issue POST '/' 경로의 라우터를 추가했습니다.

j03y14 committed:
[feat] selectIssue, selectIssuelabel query 추가

전체 이슈를 모두 가져오는 selectIssue query와 입력한 이슈에 대한 label들을 모두 가져오는 selectIssuelabel query를 추가했습니다.

j03y14 committed:
[feat] issue를 가져오는 메소드 추가

issue를 먼저 join 해서 milestone, user와 join 해서 가져온 후에 각 issue에 해당하는 label들을 가져와서 합친 후 반환합니다.

j03y14 committed:
[feat] 해당 이슈에 해당하는 label들을 가져오는 메소드 구현

query를 한 결과가 TextRow로 반환되었기 때문에 object로 바꿔주기 위해서 spread 연산자를 이용해서 convert 해준 후에 반환했습니다.

j03y14 committed:
[feat] issue controller read 메소드 구현

j03y14 committed:
[feat] issue router의 get api 구현

issue router에서 GET '/' 하면 read 하도록 했습니다.

j03y14 committed:
[feat] update issue title query 추가

입력한 issue_id와 같은 것의 title을 변경시키는 query 추가

j03y14 committed:
[feat] issue model title 변경 메소드 추가

issue model에서 issue_id에 해당하는 issue의 title을 변경하는 메소드를 구현했습니다.

j03y14 committed:
[feat] controller에 title을 변경하는 메소드 추가

title을 변경하는 메소드를 추가했습니다.

j03y14 committed:
[feat]  title을 변경하는 url 추가

PATCH /api/issue/:issueId/title로 들어오는 요청을 ㅊ처리하도록 api를 추가했습니다.

j03y14 committed:
[feat] issue에 milestone을 추가하는 api 구현

issue에 milestone을 추가하는 api를 구현했습니다.

j03y14 committed:
[feat] issue의 milestone을 삭제하는 api 구현

issue의 milestone을 삭제하는 api를 구현했습니다.
delete이지만 실제로 issue table에서는 milestone_id를 null로 바꿔주는 것으로 구현했습니다. issue에서 milestone을 지정한 것을 해제하는 개념으로 생각하면 될 것 같습니다.

j03y14 committed:
[feat] issue에 label을 추가하는 api 구현

create issue할 때 사용한 insertIssueLabel query를 재사용했습니다. insertIssueLabel가 bulk insert로 되어 있기 때문에 issue controller에서 labelId를 배열에 넣어서 issueModel의 insert 메소드에 넣었습니다.

j03y14 committed:
[feat] issue에서 label을 삭제하는 api 구현

해당 issue에서 label을 삭제하는 api를 구현했습니다.

j03y14 committed:
[feat] issue의 isopen을 수정하는 api 구현

issue의 isopen을 수정하는 api를 구현했습니다.

j03y14 committed:
[feat] issue에 assignee를 추가, 삭제하는 api를 만들었습니다.

j03y14 committed:
[feat] 해당 이슈에 해당하는 assignee를 가져오는 api 구현

해당 이슈에 해당하는 assignee를 가져오는 api 구현했습니다.

safience-mia commented on web-BE/models/issue.js:
```diff
@@ -0,0 +1,27 @@
+  insert: async ({
+    writer, title, milestone, writeTime, labelArr,
+  }) => {
+    const issueTableData = [writer, writeTime, title, milestone];
```
이슈 데이터에 is_open값이 빠진듯 합니다!
`const issueTableData = [writer, writeTime, title, is_open, milestone];`

safience-mia commented on web-BE/models/issue_label.js:
```diff
@@ -0,0 +1,12 @@
+const issueLabelModel = {
+  insert: async (issueId, labelArr) => {
+    const issueLabelData = labelArr.reduce((acc, curr) => [...acc, [issueId, curr]], []);
+    console.log('asdfsddfsef', issueLabelData);
```
ㅎㅎ 고민의 흔적이 보이는 콘솔로그,, 이거 혹시 콘솔로그 어떻게 나오는지 볼 수 있을까요? 
issueLabelDatd에 라벨 id는 포함 안되있는거죠? 쿼리에 label_id도 인자로 들어가야 하지 않을까요?🧐
`insertIssueLabel: 'INSERT INTO Issue_Label (issue_id, label_id) VALUES ?',`

j03y14 replied:
issueId가 6이고 labelArr이 [1, 2]일 때 log가 다음과 같이 나옵니다.
```javascript
// issueId=6
// labelArr=[1,2]

[ 
  [ 6, 1 ], 
  [ 6, 2 ]
]
```

j03y14 replied:
이게 한 번에 여러 행을 넣을 때 bulk insert를 한다고 하던데 query에 values가 들어가는 곳에 물음표를 하나만 두고 채connection.query를 할 때 2차원 배열로 여러 데이터를 넣으면 bulk insert가 된다고 하더라고요 그래서 위에 issueLabelData가 2차원 배열이고 2차원 배열 안에 각 배열에 issue_id랑 label_id가 들어가 있는거에요
```javascript
[
  [issue_id, label_id],
  [issue_id, label_id]
]

```

safience-mia replied:
완전 이해됐습니다 ,감사해요!
이건 리포트해서 저장해놔야겠네요, 나중에 참고할수있도록!! 
고급진 코드입니다..

safience-mia commented on web-BE/config/query.js:
```diff
@@ -20,6 +20,22 @@ module.exports = {
   insertIssueLabel: 'INSERT INTO Issue_Label (issue_id, label_id) VALUES ?',
+  selectIssueLabel:
+    'select Label.label_id, Label.`name` as label_name, Label.description, Label.color from Issue_Label '
+    + 'join Label on Issue_Label.label_id = Label.label_id WHERE Issue_Label.issue_id = ?',
```
깔끔하고 멋진 쿼리문 잘 구경하고 갑니다,,, 👍🏻 
라벨네임을 `` Label.`name` as...``으로 가져오는 이유가 있을까요? 
`` \`User\` `` 도 마찬가지로요. 변수로 가져오는걸까요?

SeunghyoKu replied:
저도 이 부분이 궁금합니다.. !

j03y14 replied:
제가 sql 쓸 때 workbench에서 미리 써보고 넣었는데 거기서 색이 다르게 나오더라고요 그래서 혹시 예약어인가 해서 저렇게 넣었습니다! 
예약어랑 같은 이름인 테이블이나 컬럼명을 쓸 때 저렇게 쓴다고 하더라고요.

근데 지금 지우고 다시 해보니까 예약어는 아닌가봐요 지우고 다시 커밋해야겠네요

safience-mia commented on web-BE/models/issue_label.js:
```diff
@@ -4,9 +4,13 @@ const sql = require('../config/query');
 const issueLabelModel = {
   insert: async (issueId, labelArr) => {
     const issueLabelData = labelArr.reduce((acc, curr) => [...acc, [issueId, curr]], []);
```
이부분은 아무래도 코드만으로 이해하기 어려우니 머지가 된 후 콘솔로그 찍어볼게요! 
issueLabelData 가 어떻게 생겼는지 보고 싶어요 ㅋㅋ

SeunghyoKu replied:
엄청난 `reduce`의 세계.. 이렇게 하면 `[issue_id, label_id]`로 반환 되는 건가요......?  🤔

j03y14 replied:
```javascript
[
  [issue_id, label_id],
  [issue_id, label_id]
]
```
이런식이에요!

safience-mia commented on web-BE/routes/api/issue/index.js:
```diff
@@ -4,6 +4,7 @@ const issueController = require('./controller');
 router.post('/', issueController.create);
 router.get('/', issueController.read);
 router.patch('/:issueId/title', issueController.updateTitle);
+router.post('/:issueId/milestone/:milestoneId', issueController.addMilestone);
```
👍🏻  깔끔하고 군더더기 없는 커밋 코드들 쭉쭉 보고 있습니다, 중간쯤에서 따봉 날리고 가여 👍🏻

slogup-mia reviewed (commented):
믿고 맡기는 재윤님의 PR 잘 보았습니다,, 감동적이네요,, 
깔끔한 코드와, 코드가 바로 읽히도록 이해를 도운 커밋 메세지도 리뷰하면서 감사할 정도였습니다!
몇가지 질문드린 부분만 해결이 되면 Approve날리겠습니다!

SeunghyoKu commented on web-BE/config/query.js:
```diff
@@ -15,4 +15,11 @@ module.exports = {
+  insertIssue:
+    'INSERT INTO Issue '
+    + '(writer_id, write_time, title, is_open, milestone_id) '
+    + 'VALUES((SELECT user_id from User WHERE username=?), ?, ?, 1, ?)',
```
쿼리문 안에 쿼리문을 작성해서 불러올 수가 있군요! 한 수 배워갑니다..

SeunghyoKu commented on web-BE/routes/api/issue/controller.js:
```diff
@@ -0,0 +1,33 @@
+    try {
+      const insertId = await issueModel.insert({
+        writer, title, milestone, writeTime, labelArr,
+      });
```
여기도 `is_open`이 빠져있는 것 같아요. 👀

j03y14 replied:
issue를 처음에 생성할 때는 무조건 open으로 넣어야 되는 것 같아서 query에 직접 1을 넣었어요!

SeunghyoKu commented on web-BE/models/issue.js:
```diff
@@ -22,6 +22,19 @@ const issueModel = {
+    try {
+      const [issueArr] = await connection.query(sql.selectIssue);
+      const labelsOfIssue = await Promise.all(
+        issueArr.map((issue) => issueLabelModel.select(issue.issue_id)),
```
`map()`을 사용해주셔서 좋은 것 같아요. map, reduce, filter와 한 몸이신.. 재윤님..

j03y14 replied:
다 찾아보면서 한거라서 부끄럽네요ㅎㅎ

safience-mia replied:
고오급 고차함수와 친하지 못해 슬픈 개린이 여기있읍니다,,

SeunghyoKu commented on web-BE/routes/api/issue/controller.js:
```diff
@@ -27,8 +27,15 @@ const issueController = {
+    const { title } = req.body;
+    try {
+      await issueModel.updateTitle(issueId, title);
+      res.sendStatus(200);
```
update가 성공적으로 완료 되면 200 status 보내줘서 확인시켜주었군요. 👀
따로 메시지는 안 보냈지만 사용자가 상태코드로 알 수 있어서 좋은 거 같아요.
저번에 말씀하셨던 상태코드로 보내는 게 이 방법이었군요.

SeunghyoKu commented on web-BE/routes/api/issue/controller.js:
```diff
@@ -46,6 +46,15 @@ const issueController = {
     }
   },
+  deleteMilestone: async (req, res, next) => {
+    const { issueId } = req.params;
```
하나를 받아오더라도 객체에 받아오는 이유가 있는지 궁금합니다. 👀

j03y14 replied:
저는 그냥 이게 더 깔끔한 느낌이 들고 나중에 파라미터를 하나를 더 추가해야될 경우에 더 변화 없이 추가할 수 있을 것 같아서 이렇게 했어요ㅎㅎ

safience-mia replied:
요거 `issueId = req.params.issueId` 대신 고급지게 쓴 할당문법이지 않을까요? 👀

SeunghyoKu commented on web-BE/config/query.js:
```diff
@@ -46,4 +46,7 @@ module.exports = {
   deleteAssignee: 'DELETE FROM Assignee WHERE issue_id=? and user_id=?',
+  selectAssignee:
+    'SELECT `User`.user_id, `User`.username, `User`.social from Assignee '
+    + 'JOIN `User` ON Assignee.user_id = `User`.user_id WHERE Assignee.issue_id=?',
```
쿼리문이 복잡한 데 뚝딱뚝딱 하신 재윤님께 1따봉 드리고 갑니다...

SeunghyoKu reviewed (approved):
매번 질문과 배움이 남는 재윤님의 코드입니다.
m:n 관계를 가지는 테이블은 어떻게 처리하면 좋을까 생각했는데, 덕분에 배운 것 같습니다.
특히 map, reduce, filter 같은 함수들을 자주 사용하셔서 배우는 게 많은 것 같아요. 👍

j03y14 committed:
[refactor] 불필요한 코드 삭제

1. is_open을 insert에서 무조건 1로 넣어주기 때문에 그냥 table의 default 값을 1로 해주고 query문에서 해당 값을 뺐습니다.

2. User와 name이 혹시 예약어인가 싶어 ``로 감싸주었지만 예약어가 아니어서 ``로 감쌌던 것들을 감싸지 않도록 수정했습니다.
